File: Uber/4/4.cpp
#include <iostream>
#include <queue>
#include <vector>

struct Node {
    Node *left, *right;
    Node(): left(nullptr), right(nullptr) { }
};

int levelWithMaximumNodes(Node* root) {
    if (!root) {
        return -1;
    }
    std::queue<Node*> NodeQueue;
    NodeQueue.push(root);
    int level = 0, maxNodeCount = -1, levelWithMaxNode = 0;
    while (!NodeQueue.empty()) {
        int NodeCount = NodeQueue.size();
        if (NodeCount > maxNodeCount) {
            maxNodeCount = NodeCount;
            levelWithMaxNode = level;
        }
        while (NodeCount--) {
            Node* tempNode = NodeQueue.front();
            NodeQueue.pop();
            if (tempNode->left) {
                NodeQueue.push(tempNode->left);
            }
            if (tempNode->right) {
                NodeQueue.push(tempNode->right);
            }
        }
        level++;
    }
    return levelWithMaxNode;
}

int main() {
    int m, n;
    std::cin >> m >> n;
    std::vector<Node*> tree;
    tree.reserve(m);
    for (int i = 0; i < m; i++) {
        tree.push_back(new Node);
    }
    while (m--) {
        int idx, l, r;
        std::cin >> idx >> l >> r;
        if (l >= 0) tree[idx]->left = tree[l];
        if (r >= 0) tree[idx]->right = tree[r];
    }
    std::cout << levelWithMaximumNodes(tree[n]) + 1 << "\n";
    return 0;
}

File: Uber/4/4.py
from queue import Queue

class newNode: 
    def __init__(self, data): 
        self.data = data
        self.left = None
        self.right = None

def levelWithMaximumNodes(root):
    if(root == None):
        return -1
    NodeQueue = Queue()
    NodeQueue.put(root)
    level = 0
    maxNodeCount = -1
    levelWithMaxNode = 0
    while not NodeQueue.empty():
        NodeCount = NodeQueue.qsize()
        if(NodeCount > maxNodeCount):
            maxNodeCount = NodeCount
            levelWithMaxNode = level
        while(NodeCount > 0):
            tempNode = NodeQueue.get()
            if tempNode.left is not None:
                NodeQueue.put(tempNode.left)
            if tempNode.right is not None:
                NodeQueue.put(tempNode.right)
            NodeCount -= 1
        level += 1
    return levelWithMaxNode

m, n = map(int, input().split())
tree = []
for x in range(m):
    tree.append(newNode(x))

for x in range(m):
    idx, l, r = map(int, input().split())
    if l >= 0:
        tree[idx].left = tree[l]
    if r >= 0:
        tree[idx].right = tree[r]
a = levelWithMaximumNodes(tree[n])
print(a + 1)
